sklearn/cross_decomposition/_pls.py

```diff
@@ -238,122 +242,41 @@ def fit(self, X, Y):
-                        raise
-                    warnings.warn(f"Y residual is constant at iteration {k}")
-                    break
-
-                self.n_iter_.append(n_iter_)
-
-            elif self.algorithm == "svd":
-                x_weights, y_weights = _get_first_singular_vectors_svd(Xk, Yk)
-
-            # inplace sign flip for consistency across solvers and archs
-            _svd_flip_1d(x_weights, y_weights)
-
-            # compute scores, i.e. the projections of X and Y
-            x_scores = np.dot(Xk, x_weights)
-            if norm_y_weights:
-                y_ss = 1
-            else:
-                y_ss = np.dot(y_weights, y_weights)
-            y_scores = np.dot(Yk, y_weights) / y_ss
+        if self.algorithm in ["nipals", "svd"]:
+            self._nipals_svd(Xk, Yk, n_samples, n_features, n_targets, n_components)
 
-            # Deflation: subtract rank-one approx to obtain Xk+1 and Yk+1
-            x_loadings = np.dot(x_scores, Xk) / np.dot(x_scores, x_scores)
-            Xk -= np.outer(x_scores, x_loadings)
+        elif self.algorithm == "dayalmacgregor":
+            self._dayal_macgregor(Xk, Yk, n_features, n_targets, n_components)
 
-            if self.deflation_mode == "canonical":
-                # regress Yk on y_score
-                y_loadings = np.dot(y_scores, Yk) / np.dot(y_scores, y_scores)
-                Yk -= np.outer(y_scores, y_loadings)
-            if self.deflation_mode == "regression":
-                # regress Yk on x_score
-                y_loadings = np.dot(x_scores, Yk) / np.dot(x_scores, x_scores)
-                Yk -= np.outer(x_scores, y_loadings)
-
-            self.x_weights_[:, k] = x_weights
-            self.y_weights_[:, k] = y_weights
-            self._x_scores[:, k] = x_scores
-            self._y_scores[:, k] = y_scores
-            self.x_loadings_[:, k] = x_loadings
-            self.y_loadings_[:, k] = y_loadings
-
-        # X was approximated as Xi . Gamma.T + X_(R+1)
-        # Xi . Gamma.T is a sum of n_components rank-1 matrices. X_(R+1) is
-        # whatever is left to fully reconstruct X, and can be 0 if X is of rank
-        # n_components.
-        # Similarly, Y was approximated as Omega . Delta.T + Y_(R+1)
-
-        # Compute transformation matrices (rotations_). See User Guide.
-        self.x_rotations_ = np.dot(
-            self.x_weights_,
-            pinv2(np.dot(self.x_loadings_.T, self.x_weights_), check_finite=False),
-        )
-        self.y_rotations_ = np.dot(
-            self.y_weights_,
-            pinv2(np.dot(self.y_loadings_.T, self.y_weights_), check_finite=False),
-        )
+        # expose the fitted attributes `x_scores_` and `y_scores_`
+        if self.deflation_mode == "regression":
+            self.x_scores_ = self._x_scores
+            if self.algorithm in ["nipals", "svd"]:
+                # Y is only deflated for NIPALS and SVD
+                self.y_scores_ = self._y_scores
```

Comment:
I see what you're going for, yeah. Implemented 👍

Reply:
All scikit-learn estimators need to pass there hyper-parameters in the constructor, and the choice of the solver is an hyper-parameter in practice.

> I see what you're going for, yeah. Implemented :+1:

I cannot see this implemented in the diff of the PR, where is this?

Can you please add a test that checks that those attributes are effectively forgotten if you you first fit with `"nipals"` and then with `"dayalmacgregor"` afterwards on the same estimator:

```python
pls = PLSRegressor(algorithm="nipals").fit(X, Y)
assert pls.y_scores_.shape == what_is_expected
assert pls.y_rotations_.shape == what_is_expected_again

pls.set_params(algorithm="dayalmacgregor").fit(X, Y)
assert not hasattr(pls, "y_scores_")
assert not hasattr(pls, "y_rotations_")
```

Reply:
sorry I was unclear - basically I rolled back my implementation because I thought this didn't apply due to the argument being passed on the constructor rather than `fit` method.

however I didn't realize Python's `set_params` usage is encouraged with estimator classes, so I can see the motivation.

this has been reimplemented with a corresponding test as well.